File: Generic_LL/QueueSinglyLL_G.java

class node<T>
{
    public T data;
    public node<T> next;

    public node(T iValue)
    {
        data = iValue;
        next = null;
    }
}

class Queue<T>
{
    private node<T> First;
    private int iCount;

    public Queue()
    {
        System.out.println("Inside Constructor !!");
        First = null;
        iCount = 0;
    }

    public void Display()
    {
        node<T> Temp = First;

        while(Temp.next != null)
        {
            System.out.print("| "+Temp.data+" | -> ");
            Temp = Temp.next;
        }
        System.out.print("| null |\n");
    }

    public int Count()
    {
        return iCount;
    }

    public void EnQueue(T iNo) // InsertLast
    {
        node<T> newn = null;
        node<T> Temp = null;

        newn = new node<T>(iNo);

        if(First == null)
        {
            First = newn;
        }
        else
        {
            Temp = First;

            while(Temp.next != null)
            {
                Temp = Temp.next;
            }
            Temp.next = newn;
        }
        iCount++;
    }

    public T DeQueue()  // DeletFirst
    {
        T iValue = null;

        if(First == null)
        {
            System.out.println("LL is Already Empty !!");
            return iValue;
        }
        else if(First.next == null)
        {
            First = null;
        }
        else
        {
            iValue = First.data;
            First = First.next;
        }
        iCount--;

        return iValue;
    }
}
class QueueSinglyLL_G
{
    public static void main(String[] args) 
    {
        Queue<Integer>Aobj = new Queue<Integer>();
        int iRet = 0;

        Aobj.EnQueue(10);
        Aobj.EnQueue(20);
        Aobj.EnQueue(30);
        Aobj.EnQueue(40);
    
        Aobj.Display();
        iRet = Aobj.Count();
        System.out.println("Number of elements in LL are :"+iRet);

        Aobj.DeQueue();

        Aobj.Display();
        iRet = Aobj.Count();
        System.out.println("Number of elements in LL are :"+iRet);

    }
    
}

File: Generic_LL/QueueSinglyLL_G.cpp
#include<iostream>
#include<stdlib.h>
using namespace std;

template <class T>
struct node
{
    T data;
    struct node *next;
};

template <class T>
class Queue
{
    private:
           struct node<T>* First;
           int iCount;
    public:
           Queue();

           void Display();
           int Count();

           void EnQueue(T iNo); // Insertlast 
           T DeQueue();         // DeleteFirst
          
};

template <class T>
Queue<T>:: Queue()
{
    cout<<"Inside counstructor !!\n";
    First = NULL;
    iCount = 0;
}

template <class T>
void Queue<T>:: Display()
{
    struct node<T>* Temp = First;
    while(Temp != NULL)
    {
        cout<<"| "<<Temp->data<<" | -> ";
        Temp = Temp->next;
    }
    cout<<"| NULL |\n";
}

template <class T>
int Queue<T>:: Count()
{
    return iCount;
}

template <class T>
void Queue<T>:: EnQueue(T iNo) // Insertlast 
{
    struct node<T>* newn = NULL;
    struct node<T>* Temp = NULL;

    newn = new struct node<T>;
    newn->data = iNo;
    newn->next = NULL;

    if(First == NULL)
    {
        First = newn;
    }
    else
    {
        Temp = First;

        while(Temp->next != NULL)
        {
            Temp = Temp->next;
        }
        Temp->next = newn;
    }
    iCount++;
}

template <class T>
T Queue<T>:: DeQueue() // DeleteFirst
{
    struct node<T>* Temp = NULL;

    if(First == NULL)
    {
        cout<<"LL is Already Empty !!\n";
        return -1;
    }
    else if(First->next == NULL)
    {
        delete First;
        First = NULL;
    }
    else
    {
        Temp = First;
        First = First->next;
        delete Temp;
    }
    iCount--;
}

int main ()
{
    Queue<int>*Aobj = new Queue<int>();
    int iRet = 0;

    Aobj->EnQueue(10);
    Aobj->EnQueue(20);
    Aobj->EnQueue(30);
    Aobj->EnQueue(40);

    Aobj->Display();
    iRet = Aobj->Count();
    cout<<"Number of elements in LL are :"<<iRet<<"\n";

    Aobj->DeQueue();

    Aobj->Display();
    iRet = Aobj->Count();
    cout<<"Number of elements in LL are :"<<iRet<<"\n";

    return 0;
}